// @cardinal/token-manager/dist/cjs/programs/timeInvalidator/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.close = exports.invalidate = exports.extendExpiration = exports.init = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const tokenManager_1 = require("../tokenManager");
const utils_1 = require("../tokenManager/utils");
const constants_1 = require("./constants");
const pda_1 = require("./pda");
const init = (connection, wallet, tokenManagerId, timeInvalidation) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const timeInvalidatorProgram = new anchor_1.Program(constants_1.TIME_INVALIDATOR_IDL, constants_1.TIME_INVALIDATOR_ADDRESS, provider);
    const [timeInvalidatorId, _timeInvalidatorBump] = yield (0, pda_1.findTimeInvalidatorAddress)(tokenManagerId);
    return [
        timeInvalidatorProgram.instruction.init({
            collector: timeInvalidation.collector || tokenManager_1.CRANK_KEY,
            paymentManager: timeInvalidation.paymentManager || tokenManager_1.PAYMENT_MANAGER_KEY,
            expiration: timeInvalidation.expiration
                ? new anchor_1.BN(timeInvalidation.expiration)
                : null,
            durationSeconds: timeInvalidation.durationSeconds
                ? new anchor_1.BN(timeInvalidation.durationSeconds)
                : null,
            extensionPaymentAmount: ((_a = timeInvalidation.extension) === null || _a === void 0 ? void 0 : _a.extensionPaymentAmount)
                ? new anchor_1.BN((_b = timeInvalidation.extension) === null || _b === void 0 ? void 0 : _b.extensionPaymentAmount)
                : null,
            extensionDurationSeconds: ((_c = timeInvalidation.extension) === null || _c === void 0 ? void 0 : _c.extensionDurationSeconds)
                ? new anchor_1.BN((_d = timeInvalidation.extension) === null || _d === void 0 ? void 0 : _d.extensionDurationSeconds)
                : null,
            extensionPaymentMint: ((_e = timeInvalidation.extension) === null || _e === void 0 ? void 0 : _e.extensionPaymentMint)
                ? (_f = timeInvalidation.extension) === null || _f === void 0 ? void 0 : _f.extensionPaymentMint
                : null,
            maxExpiration: timeInvalidation.maxExpiration
                ? new anchor_1.BN(timeInvalidation.maxExpiration)
                : null,
            disablePartialExtension: ((_g = timeInvalidation.extension) === null || _g === void 0 ? void 0 : _g.disablePartialExtension)
                ? (_h = timeInvalidation.extension) === null || _h === void 0 ? void 0 : _h.disablePartialExtension
                : null,
        }, {
            accounts: {
                tokenManager: tokenManagerId,
                timeInvalidator: timeInvalidatorId,
                issuer: wallet.publicKey,
                payer: wallet.publicKey,
                systemProgram: web3_js_1.SystemProgram.programId,
            },
        }),
        timeInvalidatorId,
    ];
});
exports.init = init;
const extendExpiration = (connection, wallet, tokenManagerId, payerTokenAccountId, timeInvalidatorId, extensionPaymentAmount, paymentAccounts) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const timeInvalidatorProgram = new anchor_1.Program(constants_1.TIME_INVALIDATOR_IDL, constants_1.TIME_INVALIDATOR_ADDRESS, provider);
    const [paymentTokenAccountId, paymentManagerTokenAccountId, remainingAccounts,] = paymentAccounts;
    return timeInvalidatorProgram.instruction.extendExpiration(new anchor_1.BN(extensionPaymentAmount), {
        accounts: {
            tokenManager: tokenManagerId,
            timeInvalidator: timeInvalidatorId,
            paymentTokenAccount: paymentTokenAccountId,
            paymentManagerTokenAccount: paymentManagerTokenAccountId,
            payer: wallet.publicKey,
            payerTokenAccount: payerTokenAccountId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
        remainingAccounts,
    });
};
exports.extendExpiration = extendExpiration;
const invalidate = (connection, wallet, mintId, tokenManagerId, tokenManagerKind, tokenManagerState, tokenManagerTokenAccountId, recipientTokenAccountId, returnAccounts) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const timeInvalidatorProgram = new anchor_1.Program(constants_1.TIME_INVALIDATOR_IDL, constants_1.TIME_INVALIDATOR_ADDRESS, provider);
    const [[timeInvalidatorId], transferAccounts] = yield Promise.all([
        (0, pda_1.findTimeInvalidatorAddress)(tokenManagerId),
        (0, utils_1.getRemainingAccountsForKind)(mintId, tokenManagerKind),
    ]);
    return timeInvalidatorProgram.instruction.invalidate({
        accounts: {
            tokenManager: tokenManagerId,
            timeInvalidator: timeInvalidatorId,
            invalidator: wallet.publicKey,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            mint: mintId,
            recipientTokenAccount: recipientTokenAccountId,
            cardinalTokenManager: tokenManager_1.TOKEN_MANAGER_ADDRESS,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
        remainingAccounts: [
            ...(tokenManagerState === tokenManager_1.TokenManagerState.Claimed
                ? transferAccounts
                : []),
            ...returnAccounts,
        ],
    });
});
exports.invalidate = invalidate;
const close = (connection, wallet, timeInvalidatorId, tokenManagerId, collector) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const timeInvalidatorProgram = new anchor_1.Program(constants_1.TIME_INVALIDATOR_IDL, constants_1.TIME_INVALIDATOR_ADDRESS, provider);
    return timeInvalidatorProgram.instruction.close({
        accounts: {
            tokenManager: tokenManagerId,
            timeInvalidator: timeInvalidatorId,
            collector: collector || tokenManager_1.CRANK_KEY,
            closer: wallet.publicKey,
        },
    });
};
exports.close = close;
