// @cardinal/token-manager/dist/cjs/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withExtendUsages = exports.withExtendExpiration = exports.withUse = exports.withInvalidate = exports.withUnissueToken = exports.withClaimToken = exports.withIssueToken = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const _1 = require(".");
const programs_1 = require("./programs");
const tokenManager_1 = require("./programs/tokenManager");
const pda_1 = require("./programs/tokenManager/pda");
const utils_1 = require("./programs/tokenManager/utils");
const utils_2 = require("./utils");
/**
 * Main method for issuing any managed token
 * Allows for optional payment, optional usages or expiration and includes a otp for private links
 * @param connection
 * @param wallet
 * @param parameters
 * @returns Transaction, public key for the created token manager and a otp if necessary for private links
 */
const withIssueToken = (transaction, connection, wallet, { claimPayment, timeInvalidation, useInvalidation, mint, amount = new anchor_1.BN(1), issuerTokenAccountId, kind = tokenManager_1.TokenManagerKind.Managed, invalidationType = tokenManager_1.InvalidationType.Return, visibility = "public", receiptOptions = undefined, customInvalidators = undefined, }) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    // init token manager
    const numInvalidator = (customInvalidators ? customInvalidators.length : 0) +
        (useInvalidation && timeInvalidation
            ? 2
            : useInvalidation || timeInvalidation
                ? 1
                : 0);
    const [tokenManagerIx, tokenManagerId] = yield programs_1.tokenManager.instruction.init(connection, wallet, mint, issuerTokenAccountId, amount, kind, invalidationType, numInvalidator);
    transaction.add(tokenManagerIx);
    //////////////////////////////
    /////// claim approver ///////
    //////////////////////////////
    let otp;
    if (claimPayment) {
        if (visibility === "private") {
            throw new Error("Private links do not currently support payment");
        }
        const [paidClaimApproverIx, paidClaimApproverId] = yield programs_1.claimApprover.instruction.init(connection, wallet, tokenManagerId, claimPayment);
        transaction.add(paidClaimApproverIx);
        transaction.add(programs_1.tokenManager.instruction.setClaimApprover(connection, wallet, tokenManagerId, paidClaimApproverId));
    }
    else if (visibility === "private") {
        otp = web3_js_1.Keypair.generate();
        transaction.add(programs_1.tokenManager.instruction.setClaimApprover(connection, wallet, tokenManagerId, otp.publicKey));
    }
    //////////////////////////////
    /////// time invalidator /////
    //////////////////////////////
    if (timeInvalidation) {
        const [timeInvalidatorIx, timeInvalidatorId] = yield programs_1.timeInvalidator.instruction.init(connection, wallet, tokenManagerId, timeInvalidation);
        transaction.add(timeInvalidatorIx);
        transaction.add(programs_1.tokenManager.instruction.addInvalidator(connection, wallet, tokenManagerId, timeInvalidatorId));
    }
    else {
        const [timeInvalidatorId] = yield programs_1.timeInvalidator.pda.findTimeInvalidatorAddress(tokenManagerId);
        const timeInvalidatorData = yield (0, utils_2.tryGetAccount)(() => programs_1.timeInvalidator.accounts.getTimeInvalidator(connection, timeInvalidatorId));
        if (timeInvalidatorData) {
            transaction.add(programs_1.timeInvalidator.instruction.close(connection, wallet, timeInvalidatorId, tokenManagerId, timeInvalidatorData.parsed.collector));
        }
    }
    //////////////////////////////
    /////////// usages ///////////
    //////////////////////////////
    if (useInvalidation) {
        const [useInvalidatorIx, useInvalidatorId] = yield programs_1.useInvalidator.instruction.init(connection, wallet, tokenManagerId, useInvalidation);
        transaction.add(useInvalidatorIx);
        transaction.add(programs_1.tokenManager.instruction.addInvalidator(connection, wallet, tokenManagerId, useInvalidatorId));
    }
    else {
        const [useInvalidatorId] = yield programs_1.useInvalidator.pda.findUseInvalidatorAddress(tokenManagerId);
        const useInvalidatorData = yield (0, utils_2.tryGetAccount)(() => programs_1.useInvalidator.accounts.getUseInvalidator(connection, useInvalidatorId));
        if (useInvalidatorData) {
            transaction.add(programs_1.useInvalidator.instruction.close(connection, wallet, useInvalidatorId, tokenManagerId, useInvalidatorData.parsed.collector));
        }
    }
    /////////////////////////////////////////
    //////////// custom invalidators ////////
    /////////////////////////////////////////
    if (customInvalidators) {
        for (const invalidator of customInvalidators) {
            transaction.add(programs_1.tokenManager.instruction.addInvalidator(connection, wallet, tokenManagerId, invalidator));
        }
    }
    if (kind === tokenManager_1.TokenManagerKind.Managed) {
        const [mintManagerIx, mintManagerId] = yield programs_1.tokenManager.instruction.creatMintManager(connection, wallet, mint);
        const mintManagerData = yield (0, utils_2.tryGetAccount)(() => programs_1.tokenManager.accounts.getMintManager(connection, mintManagerId));
        if (!mintManagerData) {
            transaction.add(mintManagerIx);
        }
    }
    // issuer
    const tokenManagerTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, mint, tokenManagerId, wallet.publicKey, true);
    transaction.add(programs_1.tokenManager.instruction.issue(connection, wallet, tokenManagerId, tokenManagerTokenAccountId, issuerTokenAccountId));
    //////////////////////////////
    //////////// index ///////////
    //////////////////////////////
    if (receiptOptions) {
        const { receiptMintKeypair } = receiptOptions;
        transaction.add(yield programs_1.tokenManager.instruction.claimReceiptMint(connection, wallet, "receipt", tokenManagerId, receiptMintKeypair.publicKey));
    }
    return [transaction, tokenManagerId, otp];
});
exports.withIssueToken = withIssueToken;
/**
 * Add claim instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param tokenManagerId
 * @param otpKeypair
 * @returns Transaction with relevent claim instructions added
 */
const withClaimToken = (transaction, connection, wallet, tokenManagerId, additionalOptions) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const [tokenManagerData, claimApproverData] = yield Promise.all([
        programs_1.tokenManager.accounts.getTokenManager(connection, tokenManagerId),
        (0, utils_2.tryGetAccount)(() => programs_1.claimApprover.accounts.getClaimApprover(connection, tokenManagerId)),
    ]);
    let claimReceiptId;
    // pay claim approver
    if (claimApproverData &&
        tokenManagerData.parsed.claimApprover &&
        tokenManagerData.parsed.claimApprover.toString() ===
            claimApproverData.pubkey.toString()) {
        const payerTokenAccountId = yield (0, _1.findAta)(claimApproverData.parsed.paymentMint, wallet.publicKey);
        [claimReceiptId] = yield programs_1.tokenManager.pda.findClaimReceiptId(tokenManagerId, wallet.publicKey);
        const paymentAccounts = yield (0, utils_1.withRemainingAccountsForPayment)(transaction, connection, wallet, claimApproverData.parsed.paymentMint, tokenManagerData.parsed.issuer, tokenManagerData.parsed.receiptMint);
        transaction.add(yield programs_1.claimApprover.instruction.pay(connection, wallet, tokenManagerId, payerTokenAccountId, paymentAccounts));
    }
    else if (tokenManagerData.parsed.claimApprover) {
        if (!(additionalOptions === null || additionalOptions === void 0 ? void 0 : additionalOptions.otpKeypair) ||
            (additionalOptions === null || additionalOptions === void 0 ? void 0 : additionalOptions.otpKeypair.publicKey.toString()) !==
                tokenManagerData.parsed.claimApprover.toString()) {
            throw new Error("Invalid OTP");
        }
        // approve claim request
        const [createClaimReceiptIx, claimReceipt] = yield programs_1.tokenManager.instruction.createClaimReceipt(connection, wallet, tokenManagerId, additionalOptions === null || additionalOptions === void 0 ? void 0 : additionalOptions.otpKeypair.publicKey);
        transaction.add(createClaimReceiptIx);
        claimReceiptId = claimReceipt;
    }
    const tokenManagerTokenAccountId = yield spl_token_1.Token.getAssociatedTokenAddress(spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID, spl_token_1.TOKEN_PROGRAM_ID, tokenManagerData.parsed.mint, tokenManagerId, true);
    const recipientTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, tokenManagerData.parsed.mint, wallet.publicKey, wallet.publicKey);
    // claim
    transaction.add(yield programs_1.tokenManager.instruction.claim(connection, wallet, tokenManagerId, tokenManagerData.parsed.kind, tokenManagerData.parsed.mint, tokenManagerTokenAccountId, recipientTokenAccountId, claimReceiptId));
    return transaction;
});
exports.withClaimToken = withClaimToken;
const withUnissueToken = (transaction, connection, wallet, mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const tokenManagerId = yield (0, pda_1.tokenManagerAddressFromMint)(connection, mintId);
    const tokenManagerTokenAccountId = yield (0, _1.findAta)(mintId, tokenManagerId, true);
    const issuerTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, mintId, wallet.publicKey, wallet.publicKey);
    return transaction.add(programs_1.tokenManager.instruction.unissue(connection, wallet, tokenManagerId, tokenManagerTokenAccountId, issuerTokenAccountId));
});
exports.withUnissueToken = withUnissueToken;
const withInvalidate = (transaction, connection, wallet, mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const tokenManagerId = yield (0, pda_1.tokenManagerAddressFromMint)(connection, mintId);
    const [[useInvalidatorId], [timeInvalidatorId]] = yield Promise.all([
        programs_1.useInvalidator.pda.findUseInvalidatorAddress(tokenManagerId),
        programs_1.timeInvalidator.pda.findTimeInvalidatorAddress(tokenManagerId),
    ]);
    const [useInvalidatorData, timeInvalidatorData, tokenManagerData] = yield Promise.all([
        (0, utils_2.tryGetAccount)(() => programs_1.useInvalidator.accounts.getUseInvalidator(connection, useInvalidatorId)),
        (0, utils_2.tryGetAccount)(() => programs_1.timeInvalidator.accounts.getTimeInvalidator(connection, timeInvalidatorId)),
        (0, utils_2.tryGetAccount)(() => programs_1.tokenManager.accounts.getTokenManager(connection, tokenManagerId)),
    ]);
    if (!tokenManagerData)
        return transaction;
    const tokenManagerTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, mintId, tokenManagerId, wallet.publicKey, true);
    const remainingAccountsForReturn = yield (0, utils_1.withRemainingAccountsForReturn)(transaction, connection, wallet, tokenManagerData);
    if (useInvalidatorData &&
        useInvalidatorData.parsed.totalUsages &&
        useInvalidatorData.parsed.usages.gte(useInvalidatorData.parsed.totalUsages)) {
        transaction.add(yield programs_1.useInvalidator.instruction.invalidate(connection, wallet, mintId, tokenManagerId, tokenManagerData.parsed.kind, tokenManagerData.parsed.state, tokenManagerTokenAccountId, tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount, remainingAccountsForReturn));
        transaction.add(programs_1.useInvalidator.instruction.close(connection, wallet, useInvalidatorId, tokenManagerId, useInvalidatorData.parsed.collector));
    }
    else if (timeInvalidatorData &&
        ((timeInvalidatorData.parsed.expiration &&
            timeInvalidatorData.parsed.expiration.lte(new anchor_1.BN(Date.now() / 1000))) ||
            (timeInvalidatorData.parsed.durationSeconds &&
                tokenManagerData.parsed.stateChangedAt
                    .add(timeInvalidatorData.parsed.durationSeconds)
                    .lte(new anchor_1.BN(Date.now() / 1000))))) {
        transaction.add(yield programs_1.timeInvalidator.instruction.invalidate(connection, wallet, mintId, tokenManagerId, tokenManagerData.parsed.kind, tokenManagerData.parsed.state, tokenManagerTokenAccountId, tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount, remainingAccountsForReturn));
        transaction.add(programs_1.timeInvalidator.instruction.close(connection, wallet, timeInvalidatorData.pubkey, timeInvalidatorData.parsed.tokenManager, timeInvalidatorData.parsed.collector));
    }
    else if (tokenManagerData.parsed.invalidators.some((inv) => inv.equals(wallet.publicKey)) ||
        tokenManagerData.parsed.invalidationType === tokenManager_1.InvalidationType.Return ||
        tokenManagerData.parsed.invalidationType === tokenManager_1.InvalidationType.Reissue) {
        transaction.add(yield programs_1.tokenManager.instruction.invalidate(connection, wallet, mintId, tokenManagerId, tokenManagerData.parsed.kind, tokenManagerData.parsed.state, tokenManagerTokenAccountId, tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount, remainingAccountsForReturn));
    }
    return transaction;
});
exports.withInvalidate = withInvalidate;
const withUse = (transaction, connection, wallet, mintId, usages, collector) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const tokenManagerId = yield (0, pda_1.tokenManagerAddressFromMint)(connection, mintId);
    const [useInvalidatorId] = yield programs_1.useInvalidator.pda.findUseInvalidatorAddress(tokenManagerId);
    const [useInvalidatorData, tokenManagerData] = yield Promise.all([
        (0, utils_2.tryGetAccount)(() => programs_1.useInvalidator.accounts.getUseInvalidator(connection, useInvalidatorId)),
        (0, utils_2.tryGetAccount)(() => programs_1.tokenManager.accounts.getTokenManager(connection, tokenManagerId)),
    ]);
    if (!useInvalidatorData) {
        // init
        const [InitTx] = yield programs_1.useInvalidator.instruction.init(connection, wallet, tokenManagerId, { collector: collector });
        transaction.add(InitTx);
    }
    if (!(tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount))
        throw new Error("Token manager has not been claimed");
    // use
    transaction.add(yield programs_1.useInvalidator.instruction.incrementUsages(connection, wallet, tokenManagerId, tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount, usages));
    if ((useInvalidatorData === null || useInvalidatorData === void 0 ? void 0 : useInvalidatorData.parsed.totalUsages) &&
        (useInvalidatorData === null || useInvalidatorData === void 0 ? void 0 : useInvalidatorData.parsed.usages.add(new anchor_1.BN(usages)).gte(useInvalidatorData === null || useInvalidatorData === void 0 ? void 0 : useInvalidatorData.parsed.totalUsages))) {
        const tokenManagerTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, mintId, tokenManagerId, wallet.publicKey, true);
        const remainingAccountsForReturn = yield (0, utils_1.withRemainingAccountsForReturn)(transaction, connection, wallet, tokenManagerData);
        transaction.add(yield programs_1.useInvalidator.instruction.invalidate(connection, wallet, mintId, tokenManagerId, tokenManagerData.parsed.kind, tokenManagerData.parsed.state, tokenManagerTokenAccountId, tokenManagerData === null || tokenManagerData === void 0 ? void 0 : tokenManagerData.parsed.recipientTokenAccount, remainingAccountsForReturn));
        transaction.add(programs_1.useInvalidator.instruction.close(connection, wallet, useInvalidatorId, tokenManagerId, useInvalidatorData.parsed.collector));
    }
    return transaction;
});
exports.withUse = withUse;
const withExtendExpiration = (transaction, connection, wallet, tokenManagerId, paymentAmount) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const [timeInvalidatorId] = yield programs_1.timeInvalidator.pda.findTimeInvalidatorAddress(tokenManagerId);
    const [timeInvalidatorData, tokenManagerData] = yield Promise.all([
        programs_1.timeInvalidator.accounts.getTimeInvalidator(connection, timeInvalidatorId),
        programs_1.tokenManager.accounts.getTokenManager(connection, tokenManagerId),
    ]);
    if (timeInvalidatorData && timeInvalidatorData.parsed.extensionPaymentMint) {
        const payerTokenAccountId = yield (0, _1.findAta)(timeInvalidatorData.parsed.extensionPaymentMint, wallet.publicKey);
        const paymentAccounts = yield (0, utils_1.withRemainingAccountsForPayment)(transaction, connection, wallet, timeInvalidatorData.parsed.extensionPaymentMint, tokenManagerData.parsed.issuer, tokenManagerData.parsed.receiptMint);
        transaction.add(programs_1.timeInvalidator.instruction.extendExpiration(connection, wallet, tokenManagerId, payerTokenAccountId, timeInvalidatorId, paymentAmount, paymentAccounts));
    }
    else {
        console.log("No payment mint");
    }
    return transaction;
});
exports.withExtendExpiration = withExtendExpiration;
const withExtendUsages = (transaction, connection, wallet, tokenManagerId, paymentAmount) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const [useInvalidatorId] = yield programs_1.useInvalidator.pda.findUseInvalidatorAddress(tokenManagerId);
    const [useInvalidatorData, tokenManagerData] = yield Promise.all([
        programs_1.useInvalidator.accounts.getUseInvalidator(connection, useInvalidatorId),
        programs_1.tokenManager.accounts.getTokenManager(connection, tokenManagerId),
    ]);
    if (useInvalidatorData && useInvalidatorData.parsed.extensionPaymentMint) {
        const payerTokenAccountId = yield (0, utils_2.withFindOrInitAssociatedTokenAccount)(transaction, connection, useInvalidatorData.parsed.extensionPaymentMint, wallet.publicKey, wallet.publicKey);
        const paymentAccounts = yield (0, utils_1.withRemainingAccountsForPayment)(transaction, connection, wallet, useInvalidatorData.parsed.extensionPaymentMint, tokenManagerData.parsed.issuer, tokenManagerData.parsed.receiptMint);
        transaction.add(programs_1.useInvalidator.instruction.extendUsages(connection, wallet, tokenManagerId, payerTokenAccountId, useInvalidatorId, paymentAmount, paymentAccounts));
    }
    return transaction;
});
exports.withExtendUsages = withExtendUsages;
