// @cardinal/namespaces-components/dist/esm/common/Button.js
import { __awaiter, __rest } from "tslib";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { lighten } from 'polished';
import { useState } from 'react';
import { LoadingSpinner } from './LoadingSpinner';
export const Button = styled.button `
  display: flex;
  align-items: center;
  gap: 5px;
  cursor: ${({ disabled }) => !disabled && 'pointer'};
  opacity: ${({ disabled }) => (disabled ? '0.5' : '1')};
  border: none;
  outline: none;
  height: 26px;
  font-size: 12px;
  mix-blend-mode: normal;
  box-shadow: ${({ boxShadow }) => boxShadow ? '0px 4px 4px rgba(0, 0, 0, 0.25)' : ''};
  border-radius: 4px;
  padding: 0 12px;
  transition: 0.2s background;
  ${({ variant = 'primary', disabled, bgColor = undefined }) => {
    return bgColor
        ? css `
          background: ${bgColor};
          color: ${getColorByBgColor(bgColor)};
          &:hover {
            background: ${!disabled && lighten(0.1, bgColor)}};
          }
        `
        : variant === 'primary'
            ? css `
          background: rgb(29, 155, 240);
          color: #fff;
          &:hover {
            background: ${!disabled && lighten(0.1, 'rgb(29, 155, 240)')}};
          }
        `
            : variant === 'secondary'
                ? css `
          background: #000;
          color: #fff;
          &:hover {
            background: ${!disabled && lighten(0.1, '#000')};
          }
        `
                : css `
          background: rgb(255, 255, 255, 0.15);
          color: #fff;
          &:hover {
            background: ${!disabled && lighten(0.05, '#000')};
          }
        `;
}}
  & > span {
    font-size: 14px;
  }
`;
export const ButtonLight = styled.div `
  border-radius: 5px;
  padding: 5px 8px;
  border: none;
  background: #eee;
  color: #777;
  cursor: pointer;
  transition: 0.1s all;
  &:hover {
    background: #ddd;
  }
`;
export const hexColor = (colorString) => {
    if (colorString.includes('#'))
        return colorString;
    const [r, g, b] = colorString
        .replace('rgb(', '')
        .replace('rgba(', '')
        .replace(')', '')
        .replace(' ', '')
        .split(',');
    return ('#' +
        [r, g, b]
            .map((x) => {
            const hex = parseInt(x || '').toString(16);
            return hex.length === 1 ? '0' + hex : hex;
        })
            .join(''));
};
export const getColorByBgColor = (bgColor) => {
    if (!bgColor) {
        return '';
    }
    return parseInt(hexColor(bgColor).replace('#', ''), 16) > 0xffffff / 2
        ? '#000'
        : '#fff';
};
export const AsyncButton = (_a) => {
    var { children, handleClick, className, color } = _a, buttonProps = __rest(_a, ["children", "handleClick", "className", "color"]);
    const [loading, setLoading] = useState(false);
    return (_jsx(Button, Object.assign({}, buttonProps, { className: className, onClick: () => __awaiter(void 0, void 0, void 0, function* () {
            try {
                setLoading(true);
                yield handleClick();
            }
            finally {
                setLoading(false);
            }
        }) }, { children: loading ? (_jsx(LoadingSpinner, { fill: color ? color : '#FFF', height: "15px" })) : (children) })));
};
export const AsyncButtonLight = ({ children, handleClick, className, color, }) => {
    const [loading, setLoading] = useState(false);
    return (_jsx(ButtonLight, Object.assign({ className: className, onClick: () => __awaiter(void 0, void 0, void 0, function* () {
            try {
                setLoading(true);
                yield handleClick();
            }
            finally {
                setLoading(false);
            }
        }) }, { children: loading ? (_jsx(LoadingSpinner, { fill: color ? color : '#FFF', height: "15px" })) : (children) })));
};
