// @cardinal/stats/dist/cjs/programs/cardinalStats/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateStatsEntry = exports.initStatsEntry = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const constants_1 = require("./constants");
const initStatsEntry = (connection, wallet, statsEntryId, name) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const statsProgram = new anchor_1.Program(constants_1.STATS_IDL, constants_1.STATS_ADDRESS, provider);
    return statsProgram.instruction.initStatsEntry({
        name: name,
    }, {
        accounts: {
            statsEntry: statsEntryId,
            authority: wallet.publicKey,
            systemProgram: web3.SystemProgram.programId,
        },
    });
};
exports.initStatsEntry = initStatsEntry;
const updateStatsEntry = (connection, wallet, statsEntryId, value) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const statsProgram = new anchor_1.Program(constants_1.STATS_IDL, constants_1.STATS_ADDRESS, provider);
    return statsProgram.instruction.updateStatsEntry(value, {
        accounts: {
            statsEntry: statsEntryId,
            authority: wallet.publicKey,
        },
    });
};
exports.updateStatsEntry = updateStatsEntry;
