// @cardinal/token-manager/dist/esm/programs/tokenManager/accounts.js
import { __awaiter } from "tslib";
import { AnchorProvider, BN, BorshAccountsCoder, Program, utils, } from "@project-serum/anchor";
import { TOKEN_MANAGER_ADDRESS, TOKEN_MANAGER_IDL } from "./constants";
export const getTokenManager = (connection, tokenManagerId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const tokenManagerProgram = new Program(TOKEN_MANAGER_IDL, TOKEN_MANAGER_ADDRESS, provider);
    const parsed = yield tokenManagerProgram.account.tokenManager.fetch(tokenManagerId);
    return {
        parsed,
        pubkey: tokenManagerId,
    };
});
export const getTokenManagers = (connection, tokenManagerIds) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const tokenManagerProgram = new Program(TOKEN_MANAGER_IDL, TOKEN_MANAGER_ADDRESS, provider);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    let tokenManagers = [];
    try {
        tokenManagers =
            yield tokenManagerProgram.account.tokenManager.fetchMultiple(tokenManagerIds);
    }
    catch (e) {
        console.log(e);
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return tokenManagers.map((tm, i) => ({
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        parsed: tm,
        pubkey: tokenManagerIds[i],
    }));
});
export const getTokenManagersByState = (connection, state) => __awaiter(void 0, void 0, void 0, function* () {
    const programAccounts = yield connection.getProgramAccounts(TOKEN_MANAGER_ADDRESS, {
        filters: state
            ? [
                {
                    memcmp: {
                        offset: 92,
                        bytes: utils.bytes.bs58.encode(new BN(state).toArrayLike(Buffer, "le", 1)),
                    },
                },
            ]
            : [],
    });
    const tokenManagerDatas = [];
    const coder = new BorshAccountsCoder(TOKEN_MANAGER_IDL);
    programAccounts.forEach((account) => {
        try {
            const tokenManagerData = coder.decode("tokenManager", account.account.data);
            if (tokenManagerData) {
                tokenManagerDatas.push(Object.assign(Object.assign({}, account), { parsed: tokenManagerData }));
            }
        }
        catch (e) {
            console.log(`Failed to decode token manager data`);
        }
    });
    return tokenManagerDatas.sort((a, b) => a.pubkey.toBase58().localeCompare(b.pubkey.toBase58()));
});
export const getMintManager = (connection, mintManagerId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const tokenManagerProgram = new Program(TOKEN_MANAGER_IDL, TOKEN_MANAGER_ADDRESS, provider);
    const parsed = yield tokenManagerProgram.account.mintManager.fetch(mintManagerId);
    return {
        parsed,
        pubkey: mintManagerId,
    };
});
export const getMintCounter = (connection, mintCounterId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const tokenManagerProgram = new Program(TOKEN_MANAGER_IDL, TOKEN_MANAGER_ADDRESS, provider);
    const parsed = yield tokenManagerProgram.account.mintCounter.fetch(mintCounterId);
    return {
        parsed,
        pubkey: mintCounterId,
    };
});
export const getTokenManagersForIssuer = (connection, issuerId) => __awaiter(void 0, void 0, void 0, function* () {
    const programAccounts = yield connection.getProgramAccounts(TOKEN_MANAGER_ADDRESS, {
        filters: [{ memcmp: { offset: 19, bytes: issuerId.toBase58() } }],
    });
    const tokenManagerDatas = [];
    const coder = new BorshAccountsCoder(TOKEN_MANAGER_IDL);
    programAccounts.forEach((account) => {
        try {
            const tokenManagerData = coder.decode("tokenManager", account.account.data);
            if (tokenManagerData) {
                tokenManagerDatas.push(Object.assign(Object.assign({}, account), { parsed: tokenManagerData }));
            }
        }
        catch (e) {
            console.log(`Failed to decode token manager data`);
        }
    });
    return tokenManagerDatas.sort((a, b) => a.pubkey.toBase58().localeCompare(b.pubkey.toBase58()));
});
