// @cardinal/token-manager/dist/esm/programs/useInvalidator/instruction.js
import { __awaiter } from "tslib";
import { AnchorProvider, BN, Program } from "@project-serum/anchor";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { SystemProgram, SYSVAR_RENT_PUBKEY } from "@solana/web3.js";
import { CRANK_KEY, PAYMENT_MANAGER_KEY, TOKEN_MANAGER_ADDRESS, TokenManagerState, } from "../tokenManager";
import { getRemainingAccountsForKind } from "../tokenManager/utils";
import { USE_INVALIDATOR_ADDRESS, USE_INVALIDATOR_IDL } from "./constants";
import { findUseInvalidatorAddress } from "./pda";
export const init = (connection, wallet, tokenManagerId, params) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g;
    const provider = new AnchorProvider(connection, wallet, {});
    const useInvalidatorProgram = new Program(USE_INVALIDATOR_IDL, USE_INVALIDATOR_ADDRESS, provider);
    const [useInvalidatorId, _useInvalidatorBump] = yield findUseInvalidatorAddress(tokenManagerId);
    return [
        useInvalidatorProgram.instruction.init({
            collector: params.collector || PAYMENT_MANAGER_KEY,
            paymentManager: params.paymentManager || CRANK_KEY,
            totalUsages: params.totalUsages ? new BN(params.totalUsages) : null,
            maxUsages: ((_a = params.extension) === null || _a === void 0 ? void 0 : _a.maxUsages)
                ? new BN((_b = params.extension) === null || _b === void 0 ? void 0 : _b.maxUsages)
                : null,
            useAuthority: params.useAuthority || null,
            extensionPaymentAmount: ((_c = params.extension) === null || _c === void 0 ? void 0 : _c.extensionPaymentAmount)
                ? new BN((_d = params.extension) === null || _d === void 0 ? void 0 : _d.extensionPaymentAmount)
                : null,
            extensionPaymentMint: ((_e = params.extension) === null || _e === void 0 ? void 0 : _e.extensionPaymentMint) || null,
            extensionUsages: ((_f = params.extension) === null || _f === void 0 ? void 0 : _f.extensionUsages)
                ? new BN((_g = params.extension) === null || _g === void 0 ? void 0 : _g.extensionUsages)
                : null,
        }, {
            accounts: {
                tokenManager: tokenManagerId,
                useInvalidator: useInvalidatorId,
                issuer: wallet.publicKey,
                payer: wallet.publicKey,
                systemProgram: SystemProgram.programId,
            },
        }),
        useInvalidatorId,
    ];
});
export const incrementUsages = (connection, wallet, tokenManagerId, recipientTokenAccountId, usages) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = new AnchorProvider(connection, wallet, {});
    const useInvalidatorProgram = new Program(USE_INVALIDATOR_IDL, USE_INVALIDATOR_ADDRESS, provider);
    const [useInvalidatorId] = yield findUseInvalidatorAddress(tokenManagerId);
    return useInvalidatorProgram.instruction.incrementUsages(new BN(usages), {
        accounts: {
            tokenManager: tokenManagerId,
            useInvalidator: useInvalidatorId,
            recipientTokenAccount: recipientTokenAccountId,
            user: wallet.publicKey,
        },
    });
});
export const invalidate = (connection, wallet, mintId, tokenManagerId, tokenManagerKind, tokenManagerState, tokenManagerTokenAccountId, recipientTokenAccountId, returnAccounts) => __awaiter(void 0, void 0, void 0, function* () {
    const provider = new AnchorProvider(connection, wallet, {});
    const useInvalidatorProgram = new Program(USE_INVALIDATOR_IDL, USE_INVALIDATOR_ADDRESS, provider);
    const [[useInvalidatorId], transferAccounts] = yield Promise.all([
        findUseInvalidatorAddress(tokenManagerId),
        getRemainingAccountsForKind(mintId, tokenManagerKind),
    ]);
    return useInvalidatorProgram.instruction.invalidate({
        accounts: {
            tokenManager: tokenManagerId,
            useInvalidator: useInvalidatorId,
            invalidator: wallet.publicKey,
            cardinalTokenManager: TOKEN_MANAGER_ADDRESS,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            tokenProgram: TOKEN_PROGRAM_ID,
            mint: mintId,
            recipientTokenAccount: recipientTokenAccountId,
            rent: SYSVAR_RENT_PUBKEY,
        },
        remainingAccounts: [
            ...(tokenManagerState === TokenManagerState.Claimed
                ? transferAccounts
                : []),
            ...returnAccounts,
        ],
    });
});
export const extendUsages = (connection, wallet, tokenManagerId, payerTokenAccountId, useInvalidatorId, extensionPaymentAmount, paymentAccounts) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const useInvalidatorProgram = new Program(USE_INVALIDATOR_IDL, USE_INVALIDATOR_ADDRESS, provider);
    const [paymentTokenAccountId, paymentManagerTokenAccountId, remainingAccounts,] = paymentAccounts;
    return useInvalidatorProgram.instruction.extendUsages(new BN(extensionPaymentAmount), {
        accounts: {
            tokenManager: tokenManagerId,
            useInvalidator: useInvalidatorId,
            paymentTokenAccount: paymentTokenAccountId,
            paymentManagerTokenAccount: paymentManagerTokenAccountId,
            payer: wallet.publicKey,
            payerTokenAccount: payerTokenAccountId,
            tokenProgram: TOKEN_PROGRAM_ID,
        },
        remainingAccounts,
    });
};
export const close = (connection, wallet, useInvalidatorId, tokenManagerId, collector) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const useInvalidatorProgram = new Program(USE_INVALIDATOR_IDL, USE_INVALIDATOR_ADDRESS, provider);
    return useInvalidatorProgram.instruction.close({
        accounts: {
            tokenManager: tokenManagerId,
            useInvalidator: useInvalidatorId,
            collector: collector || CRANK_KEY,
            closer: wallet.publicKey,
        },
    });
};
