// @cardinal/token-manager/dist/cjs/programs/tokenManager/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.invalidate = exports.claimReceiptMint = exports.closeMintManager = exports.creatMintManager = exports.createClaimReceipt = exports.claim = exports.unissue = exports.issue = exports.addInvalidator = exports.setTransferAuthority = exports.setClaimApprover = exports.init = exports.initMintCounter = void 0;
const tslib_1 = require("tslib");
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const __1 = require("../..");
const _1 = require(".");
const constants_1 = require("./constants");
const pda_1 = require("./pda");
const utils_1 = require("./utils");
const initMintCounter = (connection, wallet, mint) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [mintCounterId, _mintCounterBump] = yield (0, pda_1.findMintCounterId)(mint);
    return tokenManagerProgram.instruction.initMintCounter(mint, {
        accounts: {
            mintCounter: mintCounterId,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
});
exports.initMintCounter = initMintCounter;
const init = (connection, wallet, mint, issuerTokenAccountId, amount, kind, invalidationType, numInvalidators = 1) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [[tokenManagerId], [mintCounterId]] = yield Promise.all([
        (0, pda_1.findTokenManagerAddress)(mint),
        (0, pda_1.findMintCounterId)(mint),
    ]);
    return [
        tokenManagerProgram.instruction.init({
            numInvalidators,
            amount,
            kind,
            invalidationType,
        }, {
            accounts: {
                tokenManager: tokenManagerId,
                mintCounter: mintCounterId,
                mint: mint,
                issuer: wallet.publicKey,
                payer: wallet.publicKey,
                issuerTokenAccount: issuerTokenAccountId,
                systemProgram: web3_js_1.SystemProgram.programId,
            },
        }),
        tokenManagerId,
    ];
});
exports.init = init;
const setClaimApprover = (connection, wallet, tokenManagerId, claimApproverId) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    return tokenManagerProgram.instruction.setClaimApprover(claimApproverId, {
        accounts: {
            tokenManager: tokenManagerId,
            issuer: wallet.publicKey,
        },
    });
};
exports.setClaimApprover = setClaimApprover;
const setTransferAuthority = (connection, wallet, tokenManagerId, transferAuthorityId) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    return tokenManagerProgram.instruction.setTransferAuthority(transferAuthorityId, {
        accounts: {
            tokenManager: tokenManagerId,
            issuer: wallet.publicKey,
        },
    });
};
exports.setTransferAuthority = setTransferAuthority;
const addInvalidator = (connection, wallet, tokenManagerId, invalidatorId) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    return tokenManagerProgram.instruction.addInvalidator(invalidatorId, {
        accounts: {
            tokenManager: tokenManagerId,
            issuer: wallet.publicKey,
        },
    });
};
exports.addInvalidator = addInvalidator;
const issue = (connection, wallet, tokenManagerId, tokenManagerTokenAccountId, issuerTokenAccountId) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    return tokenManagerProgram.instruction.issue({
        accounts: {
            tokenManager: tokenManagerId,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            issuer: wallet.publicKey,
            issuerTokenAccount: issuerTokenAccountId,
            payer: wallet.publicKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
};
exports.issue = issue;
const unissue = (connection, wallet, tokenManagerId, tokenManagerTokenAccountId, issuerTokenAccountId) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    return tokenManagerProgram.instruction.unissue({
        accounts: {
            tokenManager: tokenManagerId,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            issuer: wallet.publicKey,
            issuerTokenAccount: issuerTokenAccountId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
    });
};
exports.unissue = unissue;
const claim = (connection, wallet, tokenManagerId, tokenManagerKind, mintId, tokenManagerTokenAccountId, recipientTokenAccountId, claimReceipt) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const remainingAccounts = yield (0, utils_1.getRemainingAccountsForKind)(mintId, tokenManagerKind);
    return tokenManagerProgram.instruction.claim({
        accounts: {
            tokenManager: tokenManagerId,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            mint: mintId,
            recipient: wallet.publicKey,
            recipientTokenAccount: recipientTokenAccountId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
        remainingAccounts: claimReceipt
            ? [
                ...remainingAccounts,
                { pubkey: claimReceipt, isSigner: false, isWritable: true },
            ]
            : remainingAccounts,
    });
});
exports.claim = claim;
const createClaimReceipt = (connection, wallet, tokenManagerId, claimApproverId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [claimReceiptId, _claimReceiptBump] = yield (0, pda_1.findClaimReceiptId)(tokenManagerId, wallet.publicKey);
    return [
        tokenManagerProgram.instruction.createClaimReceipt(wallet.publicKey, {
            accounts: {
                tokenManager: tokenManagerId,
                claimApprover: claimApproverId,
                claimReceipt: claimReceiptId,
                payer: wallet.publicKey,
                systemProgram: web3_js_1.SystemProgram.programId,
            },
        }),
        claimReceiptId,
    ];
});
exports.createClaimReceipt = createClaimReceipt;
const creatMintManager = (connection, wallet, mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [mintManagerId, _mintManagerBump] = yield (0, pda_1.findMintManagerId)(mintId);
    return [
        tokenManagerProgram.instruction.createMintManager({
            accounts: {
                mintManager: mintManagerId,
                mint: mintId,
                freezeAuthority: wallet.publicKey,
                payer: wallet.publicKey,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                systemProgram: web3_js_1.SystemProgram.programId,
            },
        }),
        mintManagerId,
    ];
});
exports.creatMintManager = creatMintManager;
const closeMintManager = (connection, wallet, mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [mintManagerId] = yield (0, pda_1.findMintManagerId)(mintId);
    return [
        tokenManagerProgram.instruction.closeMintManager({
            accounts: {
                mintManager: mintManagerId,
                mint: mintId,
                freezeAuthority: wallet.publicKey,
                payer: wallet.publicKey,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            },
        }),
        mintManagerId,
    ];
});
exports.closeMintManager = closeMintManager;
const claimReceiptMint = (connection, wallet, name, tokenManagerId, receiptMintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const [receiptMintMetadataId, recipientTokenAccountId, [receiptMintManagerId],] = yield Promise.all([
        mpl_token_metadata_1.Metadata.getPDA(receiptMintId),
        (0, __1.findAta)(receiptMintId, wallet.publicKey),
        (0, pda_1.findReceiptMintManagerId)(),
    ]);
    return tokenManagerProgram.instruction.claimReceiptMint(name, {
        accounts: {
            tokenManager: tokenManagerId,
            receiptMint: receiptMintId,
            receiptMintMetadata: receiptMintMetadataId,
            recipientTokenAccount: recipientTokenAccountId,
            issuer: wallet.publicKey,
            payer: wallet.publicKey,
            receiptMintManager: receiptMintManagerId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            associatedToken: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
            tokenMetadataProgram: mpl_token_metadata_1.MetadataProgram.PUBKEY,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
    });
});
exports.claimReceiptMint = claimReceiptMint;
const invalidate = (connection, wallet, mintId, tokenManagerId, tokenManagerKind, tokenManagerState, tokenManagerTokenAccountId, recipientTokenAccountId, returnAccounts) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const tokenManagerProgram = new anchor_1.Program(constants_1.TOKEN_MANAGER_IDL, constants_1.TOKEN_MANAGER_ADDRESS, provider);
    const transferAccounts = yield (0, utils_1.getRemainingAccountsForKind)(mintId, tokenManagerKind);
    return tokenManagerProgram.instruction.invalidate({
        accounts: {
            tokenManager: tokenManagerId,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            mint: mintId,
            recipientTokenAccount: recipientTokenAccountId,
            invalidator: wallet.publicKey,
            collector: _1.CRANK_KEY,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
        remainingAccounts: [
            ...(tokenManagerState === _1.TokenManagerState.Claimed
                ? transferAccounts
                : []),
            ...returnAccounts,
        ],
    });
});
exports.invalidate = invalidate;
