// @cardinal/token-manager/dist/cjs/programs/tokenManager/pda.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findReceiptMintManagerId = exports.findMintCounterId = exports.findMintManagerId = exports.findTransferReceiptId = exports.findClaimReceiptId = exports.findTokenManagerAddress = exports.tokenManagerAddressFromMint = exports.tryTokenManagerAddressFromMint = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const web3_js_1 = require("@solana/web3.js");
const _1 = require(".");
const constants_1 = require("./constants");
/**
 * Finds the token manager address for a given mint
 * @returns
 */
const tryTokenManagerAddressFromMint = (connection, mint) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    try {
        const tokenManagerId = yield (0, exports.tokenManagerAddressFromMint)(connection, mint);
        return tokenManagerId;
    }
    catch (e) {
        return null;
    }
});
exports.tryTokenManagerAddressFromMint = tryTokenManagerAddressFromMint;
/**
 * Finds the token manager address for a given mint
 * @returns
 */
const tokenManagerAddressFromMint = (_connection, mint) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    const [tokenManagerId] = yield (0, exports.findTokenManagerAddress)(mint);
    return tokenManagerId;
});
exports.tokenManagerAddressFromMint = tokenManagerAddressFromMint;
/**
 * Finds the token manager address for a given mint and mint counter
 * @returns
 */
const findTokenManagerAddress = (mint) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return yield web3_js_1.PublicKey.findProgramAddress([anchor_1.utils.bytes.utf8.encode(constants_1.TOKEN_MANAGER_SEED), mint.toBuffer()], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findTokenManagerAddress = findTokenManagerAddress;
/**
 * Finds the claim receipt id.
 * @returns
 */
const findClaimReceiptId = (tokenManagerKey, recipientKey) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return web3_js_1.PublicKey.findProgramAddress([
        anchor_1.utils.bytes.utf8.encode(constants_1.CLAIM_RECEIPT_SEED),
        tokenManagerKey.toBuffer(),
        recipientKey.toBuffer(),
    ], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findClaimReceiptId = findClaimReceiptId;
/**
 * Finds the transfer receipt id.
 * @returns
 */
const findTransferReceiptId = (tokenManagerKey, recipientKey) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return web3_js_1.PublicKey.findProgramAddress([
        anchor_1.utils.bytes.utf8.encode(_1.TRANSFER_RECEIPT_SEED),
        tokenManagerKey.toBuffer(),
        recipientKey.toBuffer(),
    ], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findTransferReceiptId = findTransferReceiptId;
/**
 * Finds the mint manager id.
 * @returns
 */
const findMintManagerId = (mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return web3_js_1.PublicKey.findProgramAddress([anchor_1.utils.bytes.utf8.encode(_1.MINT_MANAGER_SEED), mintId.toBuffer()], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findMintManagerId = findMintManagerId;
/**
 * Finds the mint counter id.
 * @returns
 */
const findMintCounterId = (mintId) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return web3_js_1.PublicKey.findProgramAddress([anchor_1.utils.bytes.utf8.encode(_1.MINT_COUNTER_SEED), mintId.toBuffer()], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findMintCounterId = findMintCounterId;
/**
 * Finds the receipt mint manager id.
 * @returns
 */
const findReceiptMintManagerId = () => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    return web3_js_1.PublicKey.findProgramAddress([anchor_1.utils.bytes.utf8.encode(_1.RECEIPT_MINT_MANAGER_SEED)], constants_1.TOKEN_MANAGER_ADDRESS);
});
exports.findReceiptMintManagerId = findReceiptMintManagerId;
