// @cardinal/token-manager/dist/cjs/programs/tokenManager/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withRemainingAccountsForReturn = exports.withRemainingAccountsForPayment = exports.getRemainingAccountsForKind = void 0;
const tslib_1 = require("tslib");
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const __1 = require("../..");
const _1 = require(".");
const pda_1 = require("./pda");
const getRemainingAccountsForKind = (mintId, tokenManagerKind) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    if (tokenManagerKind === _1.TokenManagerKind.Managed) {
        const [mintManagerId] = yield (0, pda_1.findMintManagerId)(mintId);
        return [
            {
                pubkey: mintManagerId,
                isSigner: false,
                isWritable: true,
            },
        ];
    }
    else if (tokenManagerKind === _1.TokenManagerKind.Edition) {
        const editionId = yield mpl_token_metadata_1.Edition.getPDA(mintId);
        return [
            {
                pubkey: editionId,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: mpl_token_metadata_1.MetadataProgram.PUBKEY,
                isSigner: false,
                isWritable: false,
            },
        ];
    }
    else {
        return [];
    }
});
exports.getRemainingAccountsForKind = getRemainingAccountsForKind;
const withRemainingAccountsForPayment = (transaction, connection, wallet, paymentMint, issuerId, receiptMint, allowOwnerOffCurve = true) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    var _a;
    if (receiptMint) {
        const receiptMintLargestAccount = yield connection.getTokenLargestAccounts(receiptMint);
        // get holder of receipt mint
        const receiptTokenAccountId = (_a = receiptMintLargestAccount.value[0]) === null || _a === void 0 ? void 0 : _a.address;
        if (!receiptTokenAccountId)
            throw new Error("No token accounts found");
        const receiptMintToken = new spl_token_1.Token(connection, receiptMint, spl_token_1.TOKEN_PROGRAM_ID, web3_js_1.Keypair.generate());
        const receiptTokenAccount = yield receiptMintToken.getAccountInfo(receiptTokenAccountId);
        // get ATA for this mint of receipt mint holder
        const [returnTokenAccountId, paymentManagerTokenAccountId] = yield Promise.all([
            receiptTokenAccount.owner.equals(wallet.publicKey)
                ? yield (0, __1.findAta)(paymentMint, receiptTokenAccount.owner, allowOwnerOffCurve)
                : yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, paymentMint, receiptTokenAccount.owner, wallet.publicKey, allowOwnerOffCurve),
            yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, paymentMint, _1.PAYMENT_MANAGER_KEY, wallet.publicKey, true),
        ]);
        return [
            returnTokenAccountId,
            paymentManagerTokenAccountId,
            [
                {
                    pubkey: receiptTokenAccountId,
                    isSigner: false,
                    isWritable: true,
                },
            ],
        ];
    }
    else {
        const [issuerTokenAccountId, paymentManagerTokenAccountId] = yield Promise.all([
            issuerId.equals(wallet.publicKey)
                ? yield (0, __1.findAta)(paymentMint, issuerId, allowOwnerOffCurve)
                : yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, paymentMint, issuerId, wallet.publicKey, allowOwnerOffCurve),
            yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, paymentMint, _1.PAYMENT_MANAGER_KEY, wallet.publicKey, true),
        ]);
        return [issuerTokenAccountId, paymentManagerTokenAccountId, []];
    }
});
exports.withRemainingAccountsForPayment = withRemainingAccountsForPayment;
const withRemainingAccountsForReturn = (transaction, connection, wallet, tokenManagerData, allowOwnerOffCurve = true) => (0, tslib_1.__awaiter)(void 0, void 0, void 0, function* () {
    var _b;
    const { issuer, mint, invalidationType, receiptMint, state } = tokenManagerData.parsed;
    if (invalidationType === _1.InvalidationType.Return ||
        state === _1.TokenManagerState.Issued) {
        if (receiptMint) {
            const receiptMintLargestAccount = yield connection.getTokenLargestAccounts(receiptMint);
            // get holder of receipt mint
            const receiptTokenAccountId = (_b = receiptMintLargestAccount.value[0]) === null || _b === void 0 ? void 0 : _b.address;
            if (!receiptTokenAccountId)
                throw new Error("No token accounts found");
            const receiptMintToken = new spl_token_1.Token(connection, receiptMint, spl_token_1.TOKEN_PROGRAM_ID, web3_js_1.Keypair.generate());
            const receiptTokenAccount = yield receiptMintToken.getAccountInfo(receiptTokenAccountId);
            // get ATA for this mint of receipt mint holder
            const returnTokenAccountId = yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, mint, receiptTokenAccount.owner, wallet.publicKey, allowOwnerOffCurve);
            return [
                {
                    pubkey: returnTokenAccountId,
                    isSigner: false,
                    isWritable: true,
                },
                {
                    pubkey: receiptTokenAccountId,
                    isSigner: false,
                    isWritable: true,
                },
            ];
        }
        else {
            const issuerTokenAccountId = yield (0, __1.withFindOrInitAssociatedTokenAccount)(transaction, connection, mint, issuer, wallet.publicKey, allowOwnerOffCurve);
            return [
                {
                    pubkey: issuerTokenAccountId,
                    isSigner: false,
                    isWritable: true,
                },
            ];
        }
    }
    else {
        return [];
    }
});
exports.withRemainingAccountsForReturn = withRemainingAccountsForReturn;
