// @saberhq/use-solana/dist/cjs/adapters/readonly/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadonlyAdapter = void 0;
const web3_js_1 = require("@solana/web3.js");
const eventemitter3_1 = require("eventemitter3");
class ReadonlyAdapter extends eventemitter3_1.EventEmitter {
    constructor() {
        var _a;
        super();
        this._publicKey = null;
        this.connect = () => {
            this.emit("connect", this._publicKey);
            return Promise.resolve();
        };
        const localPubkey = (_a = process.env.REACT_APP_LOCAL_PUBKEY) !== null && _a !== void 0 ? _a : process.env.LOCAL_PUBKEY;
        if (!localPubkey) {
            console.warn("LOCAL_PUBKEY not set for readonly provider");
        }
        else {
            this._publicKey = new web3_js_1.PublicKey(localPubkey);
        }
    }
    get connected() {
        return true;
    }
    get autoApprove() {
        return false;
    }
    get publicKey() {
        return this._publicKey;
    }
    signAllTransactions(_transactions) {
        throw new Error("readonly adapter cannot sign transactions");
    }
    signTransaction(_transaction) {
        throw new Error("readonly adapter cannot sign transactions");
    }
    disconnect() {
        this.emit("disconnect");
    }
}
exports.ReadonlyAdapter = ReadonlyAdapter;
