// @saberhq/use-solana/dist/cjs/adapters/solana.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaWalletAdapter = void 0;
const tslib_1 = require("tslib");
class SolanaWalletAdapter {
    constructor(adapter) {
        this.adapter = adapter;
        this.connect = () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.adapter.connect();
        });
    }
    get connected() {
        return this.adapter.connected;
    }
    get autoApprove() {
        return false;
    }
    signAllTransactions(transactions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.adapter.signAllTransactions(transactions);
        });
    }
    get publicKey() {
        return this.adapter.publicKey;
    }
    signTransaction(transaction) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.adapter) {
                return transaction;
            }
            return this.adapter.signTransaction(transaction);
        });
    }
    disconnect() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.adapter.disconnect();
        });
    }
    on(event, fn) {
        this.adapter.on(event, fn);
    }
}
exports.SolanaWalletAdapter = SolanaWalletAdapter;
