// @cardinal/stats/dist/esm/programs/cardinalStats/instruction.js
import { AnchorProvider, Program } from "@project-serum/anchor";
import * as web3 from "@solana/web3.js";
import { STATS_ADDRESS, STATS_IDL } from "./constants";
export const initStatsEntry = (connection, wallet, statsEntryId, name) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const statsProgram = new Program(STATS_IDL, STATS_ADDRESS, provider);
    return statsProgram.instruction.initStatsEntry({
        name: name,
    }, {
        accounts: {
            statsEntry: statsEntryId,
            authority: wallet.publicKey,
            systemProgram: web3.SystemProgram.programId,
        },
    });
};
export const updateStatsEntry = (connection, wallet, statsEntryId, value) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const statsProgram = new Program(STATS_IDL, STATS_ADDRESS, provider);
    return statsProgram.instruction.updateStatsEntry(value, {
        accounts: {
            statsEntry: statsEntryId,
            authority: wallet.publicKey,
        },
    });
};
