// @cardinal/namespaces/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLAIM_REQUEST_SEED = exports.REVERSE_ENTRY_SEED = exports.ENTRY_SEED = exports.NAMESPACE_SEED = exports.GLOBAL_CONTEXT_SEED = exports.GLOBAL_RENTAL_PERCENTAGE = exports.IDENTITIES = exports.DEFAULT_PAYMENT_MANAGER = exports.NAMESPACES_IDL = exports.METADATA_CONFIG_SEED = exports.NAMESPACES_PROGRAM_ID = void 0;
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const NAMESPACES_TYPES = tslib_1.__importStar(require("./idl/cardinal_namespaces"));
exports.NAMESPACES_PROGRAM_ID = new web3_js_1.PublicKey("nameXpT2PwZ2iA6DTNYTotTmiMYusBCYqwBLN2QgF4w");
exports.METADATA_CONFIG_SEED = "metadata-config";
exports.NAMESPACES_IDL = NAMESPACES_TYPES.IDL;
exports.DEFAULT_PAYMENT_MANAGER = "cardinal";
exports.IDENTITIES = ["twitter", "discord"];
exports.GLOBAL_RENTAL_PERCENTAGE = 0.2;
exports.GLOBAL_CONTEXT_SEED = "context";
exports.NAMESPACE_SEED = "namespace";
exports.ENTRY_SEED = "entry";
exports.REVERSE_ENTRY_SEED = "reverse-entry";
exports.CLAIM_REQUEST_SEED = "rent-request";
