// @cardinal/namespaces-components/dist/cjs/modal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Modal = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const darken_1 = tslib_1.__importDefault(require("polished/lib/color/darken"));
const react_1 = require("react");
const icons_1 = require("./icons");
const Modal = ({ children, isOpen, onDismiss, darkenOverlay = true, hideCloseButton = false, }) => {
    const [mounted, setMounted] = (0, react_1.useState)(true);
    (0, react_1.useEffect)(() => {
        !isOpen
            ? setTimeout(() => {
                setMounted(false);
            }, 200)
            : setMounted(true);
    }, [isOpen]);
    return ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsx)(StyledDialogOverlay, Object.assign({ isOpen: isOpen, darkenOverlay: darkenOverlay, onClick: (e) => {
                e.stopPropagation();
                e.preventDefault();
                onDismiss();
            } }, { children: (0, jsx_runtime_1.jsxs)(ModalWrapper, Object.assign({ onClick: (e) => {
                    e.stopPropagation();
                    e.preventDefault();
                } }, { children: [(0, jsx_runtime_1.jsxs)(TopArea, { children: [(0, jsx_runtime_1.jsx)("div", {}), hideCloseButton ? ((0, jsx_runtime_1.jsx)("div", {})) : ((0, jsx_runtime_1.jsx)(ButtonIcon, Object.assign({ onClick: (e) => {
                                    e.stopPropagation();
                                    e.preventDefault();
                                    onDismiss();
                                } }, { children: (0, jsx_runtime_1.jsx)(icons_1.CloseIcon, {}) })))] }), mounted && children] })) })) }));
};
exports.Modal = Modal;
const TopArea = styled_1.default.div `
  padding: 12px 16px 0px 16px;
  top: 12px;
  left: 16px;
  right: 16px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;
const ButtonIcon = styled_1.default.div `
  flex: 0 0 24px;
  cursor: pointer;
  color: #ccd2e3;
  &:hover {
    color: ${(0, darken_1.default)(0.1, '#ccd2e3')};
  }
  transition: 0.1s ease;
`;
const ModalWrapper = styled_1.default.div `
  * {
    box-sizing: border-box;
  }
  font-family: SFHello, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
    Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji',
    'Segoe UI Symbol';
  position: relative;
  box-shadow: 0px 4px 16px rgba(207, 207, 207, 0.25);
  width: 100%;
  max-width: 560px;
  border-radius: 8px;
  background: #fff;
  color: #696969;
  font-weight: normal;
  font-size: 12px;
  line-height: 15px;
  letter-spacing: -0.02em;
  color: #696969;
  margin: 10vh auto;
`;
const StyledDialogOverlay = styled_1.default.div `
  transition: 0.2s all;
  width: 100vw;
  height: 100vh;
  position: absolute;
  visibility: ${({ isOpen }) => (isOpen ? 'visible' : 'collapse')};
  opacity: ${({ isOpen }) => (isOpen ? '1' : '0')};
  background: ${({ darkenOverlay }) => darkenOverlay ? 'rgba(0, 0, 0, 0.55)' : 'none'};
  z-index: 1000;
`;
